test/berlim_web/controllers/admin_controller_test.exs

```diff
@@ -1,87 +1,78 @@
 defmodule BerlimWeb.AdminControllerTest do
   use BerlimWeb.ConnCase
-  use BerlimWeb.Helpers.AuthHelper
 
   import Berlim.Factory
+  alias Berlim.InternalAccounts.Admin
 
   @create_attrs params_for(:admin)
   @update_attrs %{name: "Lionel Ritchie"}
   @invalid_attrs %{name: nil, email: nil}
 
-  describe "GET /index, when user is an admin" do
-    setup [:authenticate_admin]
+  setup %{conn: conn} do
+    {:ok, conn: put_req_header(conn, "accept", "application/json")}
+  end
 
+  describe "GET /index" do
     test "lists all admins", %{conn: conn} do
       conn = get(conn, Routes.admin_path(conn, :index))
-      assert html_response(conn, 200) =~ "Admins"
+      assert json_response(conn, 200)["data"] == []
     end
   end
 
-  describe "GET /new, when user is an admin" do
-    setup [:authenticate_admin]
-
-    test "renders form", %{conn: conn} do
-      conn = get(conn, Routes.admin_path(conn, :new))
-      assert html_response(conn, 200) =~ "Novo Admin"
-    end
-  end
+  describe "POST /create" do
+    test "renders admin show when data is valid", %{conn: conn} do
+      conn = post(conn, Routes.admin_path(conn, :create), admin: @create_attrs)
+      assert %{"id" => id} = json_response(conn, 201)["data"]
 
-  describe "POST /create, when user is an admin" do
-    setup [:authenticate_admin]
+      conn = get(conn, Routes.admin_path(conn, :show, id))
 
-    test "redirects to index when data is valid", %{conn: conn} do
-      conn = post(conn, Routes.admin_path(conn, :create), admin: @create_attrs)
-      assert redirected_to(conn) == Routes.admin_path(conn, :index)
+      assert %{
+               "id" => id,
```

Comment:
this `id` variable is unused, I recommend you setup credo on overcommit to avoid these issues

Reply:
como assim não é usado?

Reply:
vc está fazendo um assign do id retornado na variável id, pattern matching

Reply:
Interessante, esse teste foi baseado no scaffold gerado com gen json. Pode dar um exemplo de como ficaria melhor escrito esse teste? Somente com um assert validando status 201?

Reply:
Veja direito o scaffold, provavelmente tem mais alguma coisa além disso.

Se vc quer garantir que ao chamar a rota show retorne o mesmo id do registro criado precisa fazer:

```elixir
%{"id" => ^id, name => "John Doe"} = json_response(conn, 200)["data"]
```

Mas não entendo porque fazer 2 requests para testar o `create`, para mim basta testar que o response é 201 e renderiza a view que vc chamou lá no seu controller.

Exemplo:

```elixir
assert json_response(conn, 201) == render_json(BookView, "index.json", %{books: books})
```

Esse `render_json` não é nativo do phoenix:

```elixir
defmodule ApolloWeb.ConnCaseHelper do
  @moduledoc """
  Utility methods for tests
  """

  def render_json(view, template, assigns) do
    template
    |> view.render(assigns)
    |> format_json
  end

  def format_json(data) do
    data
    |> Poison.encode!()
    |> Poison.decode!()
  end
end
```

Reply:
feito! thx!